### ISSUE TITLE
Offline configuration for buildPythonApplication

### ISSUE THREAD

**Issue by W1M0R** [DavHau/mach-nix#237]
I use mach-nix.buildPythonApplication to install a python cli utility (called locust) from GitHub. Whenever I enter the nix-shell, the following check is performed by mach-nix:

```bash
~/proj $ nix-shell
trace:
 automatically detected requirements of locust:1.4.3 20.03post-git:
gevent>=20.9.0
flask>=1.1.2
Werkzeug>=1.0.1
requests>=2.9.1
msgpack>=0.6.2
pyzmq>=16.0.2
geventhttpclient>=1.4.4
ConfigArgParse>=1.0
psutil>=5.6.7
Flask-BasicAuth>=0.2.0

[nix-shell:~/proj]$
```

This check happens every time when I enter the nix-shell. If I am disconnected from the internet, then nix-shell fails, because of this check, that evidently requires an internet connection.

Is this a bug, or a misconfiguration of mach-nix on my part?

Here are the relevant nix derivations:

locust/default.nix:
```nix
{ buildPythonApplication, fetchFromGitHub, lib }:

buildPythonApplication rec {
  pname = "locust";
  version = "1.4.3";

  src = fetchFromGitHub {
    owner = "locustio";
    repo = "locust";
    rev = "${version}";
    sha256 = "06nxfa20234yyy1m8kyzdqblam2ym05j7hkghhj7ga5awy69cb2z";
  };

  meta = with lib; {
    license = licenses.mit;
    homepage = "https://github.com/locustio/locust";
    description = "";
    maintainers = with maintainers; [ locustio ];
  };
}
```

mach-nix/default.nix:
```nix
{ pkgs, python, fetchFromGitHub }:

let
    version = "3.1.1";

    src = fetchFromGitHub {
        owner = "DavHau";
        repo = "mach-nix";
        rev = "${version}";
        sha256 = "06bxmz7x1fnjx2d0gb63fdm0jci2mcdnpa3isbhjr92z1l2a7hl1";
    };
    
    mach-nix = import src {
        inherit pkgs;
        inherit python;
    };

in mach-nix
```

shared.nix:
```nix
{ sources ? import ./sources.nix }:
let
  pkgs = import sources.nixpkgs { };
 
  python = pkgs.python38;
  python_ver = "python38";

  mach-nix = import ./mach-nix/default.nix {
    inherit pkgs;
    inherit (pkgs) fetchFromGitHub;
    python = python_ver;
  };

  locust = import ./locust/default.nix {
    inherit (pkgs) lib fetchFromGitHub;
    inherit (mach-nix) buildPythonApplication;
  };

in
{
  inherit pkgs;

  devTools = {
    inherit python;
    inherit locust;
  };

}
```

shell.nix:
```nix
{ project ? import ./shared { }
}:

project.pkgs.mkShell {
  buildInputs = builtins.attrValues project.devTools;
}
```

**Comment by DavHau** [comment]
Hey, thanks for reporting this.
With the current stable version of nix, nix evaluations itself are not cached, only store outputs are. nixFlakes fixes that to some degree (https://www.tweag.io/blog/2020-06-25-eval-cache/).

Any other shell.nix will also be re-evaluated when you execute nix-shell. Usually those just don't print anything to stdout, so you might not notice it. Another reason why this might be more noticable with mach-nix is because it does import from derivation, which slows down evaluation.

The expression should not  require online access though, as all your fetching is pure. Which request fails in your case? What is the error?

BTW, if you install direnv + nix-direnv, it will cache the nix evaluations and offline use shouldn't be a problem.

**Comment by W1M0R** [comment]
> BTW, if you install direnv + nix-direnv, it will cache the nix evaluations and offline use shouldn't be a problem.

Thanks for the tip!

> The expression should not require online access though, as all your fetching is pure. Which request fails in your case? What is the error?

I tried to duplicate the error now upon your request, by manually disconnecting from the network...and no error occurred this time. To force the error, I did a nix-collect-garbage followed by nix-shell, and even though this was not the error, it was the same type of error:

```
warning: unable to download ...
```
I think the error mentioned something about pypi-deps-db.

I'm sorry for wasting your time. Clearly the error was a misconfiguration on my side (maybe caused by a nix garbage collect). Thanks for your response.
